### ISSUE TITLE
How to decrease false positives ?

### ISSUE THREAD

**Issue by Zegorax** [ultralytics/yolov5#666]
## ❔Question

Recently YOLO started to give me a lot of incorrect results and I thought that removing the annotation and adding the image (without .txt, blank txt or with other correct annotations) to the training set would make it learn there is nothing there, but it is not working.

What is the best way to decrease false positive rate ? How would you do it ?

**Comment by glenn-jocher** [comment]
@Zegorax you control your own FP rate by setting --conf to whatever you want at inference time. Precision is completely up to you, it's not a property of the trained model.

**Comment by Zegorax** [comment]
@glenn-jocher Hello Glenn, thank you for answering me.

I'm trying to improve the detection, because those false positive have sometimes a high confidence level.

I'm wondering what I can do to improve the model and make it not detect / predict those false positives

**Comment by glenn-jocher** [comment]
@Zegorax then you should use a larger model, more training data, increase resolution, try different augmentation, vary hyperparameters, etc.

Nothing new here, just the usual suspects for improving results.

**Comment by Zegorax** [comment]
@glenn-jocher Thank you. I improved the training set a little bit with more data. What I was wondering is the following : Is there a way to give YOLO an image without annotation and make him learn that there's nothing on this image ?

**Comment by Ownmarc** [comment]
@Zegorax yes, no annotation file is needed when there is nothing in the image. The image will still be used to train on as any other image in the train directory. I personally prefer to create an empty annotation file just so my image count always match my annotation file count.

**Comment by Zegorax** [comment]
@Ownmarc That's perfect then, it's what I've been doing the whole time. Thank you very much!

**Comment by wwdok** [comment]
> @Zegorax yes, no annotation file is needed when there is nothing in the image. The image will still be used to train on as any other image in the train directory. I personally prefer to create an empty annotation file just so my image count always match my annotation file count.

Hi, What do you think the ratio of the the dataset with objects and dataset without objects should be ？

**Comment by mujinbao** [comment]
@Zegorax @wwdok 
My own training set has 8000+ positive image, and the empty target image are about 2000 long. I think the effect is good. At least, I see a lot less of that stupid false postive.
I think if too many empty target images are added, it will make the model difficult to converge. In addition, the false positive which is very similar to the detected object is still not improved. I guess it is because the quality of my dataset is not high, the more detailed features in the original image no longer exist. 
So if anyone has a good idea on the super parameters or network parameters, please also @ me. (￣▽￣)"

**Comment by macqueen09** [comment]
I constructed an offline data augmentation. Let the falsely detected image be used as the background, and the target image is scaled to the background. 1 to 1 ratio training, due to various reasons, the accuracy rate is reduced, I will continue to try different ratios and scaling, and directly add negative samples to compare.

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!
